aggregator/leader_flush_mgr.go

```diff
@@ -0,0 +1,341 @@
+		flushers    []PeriodicFlusher
+		waitFor     = mgr.checkEvery
+	)
+	mgr.Lock()
+	numFlushTimes := mgr.flushTimes.Len()
+	mgr.metrics.queueSize.Update(float64(numFlushTimes))
+	if numFlushTimes > 0 {
+		earliestFlush := mgr.flushTimes.Min()
+		if nowNanos := mgr.nowNanos(); nowNanos >= earliestFlush.timeNanos {
+			// NB(xichen): make a shallow copy of the flushers inside the lock
+			// and use the snapshot for flushing below.
+			shouldFlush = true
+			bucketIdx := earliestFlush.bucketIdx
+			duration = buckets[bucketIdx].duration
+			flushers = buckets[bucketIdx].flushers
+			nextFlushMetadata := flushMetadata{
+				timeNanos: earliestFlush.timeNanos + int64(buckets[bucketIdx].interval),
+				bucketIdx: bucketIdx,
+			}
+			mgr.flushTimes.Pop()
+			mgr.flushTimes.Push(nextFlushMetadata)
+			mgr.flushedSincePersist = true
+		} else {
+			// NB(xichen): don't oversleep if the next flush is about to happen.
+			timeToNextFlush := time.Duration(earliestFlush.timeNanos - nowNanos)
+			if timeToNextFlush < waitFor {
+				waitFor = timeToNextFlush
+			}
+		}
+	}
+
+	var (
+		now           = mgr.nowFn()
+		shouldPersist = false
+		flushTimes    schema.ShardSetFlushTimes
+	)
+	durationSinceLastPersist := now.Sub(mgr.lastPersistAt)
+	if mgr.flushedSincePersist && durationSinceLastPersist >= mgr.flushTimesPersistEvery {
+		shouldPersist = true
+		mgr.lastPersistAt = now
+		mgr.flushedSincePersist = false
+		flushTimes = mgr.prepareFlushTimesWithLock(buckets)
+	}
+	mgr.Unlock()
+
+	if !shouldFlush && !shouldPersist {
+		return nil, waitFor
+	}
+	mgr.flushTask.shouldFlush = shouldFlush
+	mgr.flushTask.duration = duration
+	mgr.flushTask.flushers = flushers
+	mgr.flushTask.shouldPersist = shouldPersist
+	mgr.flushTask.flushTimes = &flushTimes
+	return mgr.flushTask, 0
+}
+
+// NB(xichen): if the current instance is a leader, we need to update the flush
+// times heap for the flush goroutine to pick it up.
+func (mgr *leaderFlushManager) OnBucketAdded(
+	bucketIdx int,
+	bucket *flushBucket,
+) {
+	mgr.Lock()
+	mgr.enqueueBucketWithLock(bucketIdx, bucket)
+	mgr.Unlock()
+}
+
+// NB(xichen): leader flush manager can always lead.
+func (mgr *leaderFlushManager) CanLead() bool { return true }
+
+func (mgr *leaderFlushManager) enqueueBucketWithLock(
+	bucketIdx int,
+	bucket *flushBucket,
+) {
+	nowNanos := mgr.nowNanos()
+	flushInterval := bucket.interval
+	nextFlushNanos := nowNanos + int64(flushInterval)
+	if mgr.jitterEnabled {
+		jitterNanos := int64(mgr.rand.Int63n(int64(flushInterval)))
+		nextFlushNanos = nowNanos + jitterNanos
+	}
+	newFlushMetadata := flushMetadata{
+		timeNanos: nextFlushNanos,
+		bucketIdx: bucketIdx,
+	}
+	mgr.flushTimes.Push(newFlushMetadata)
+}
+
+func (mgr *leaderFlushManager) prepareFlushTimesWithLock(
+	buckets []*flushBucket,
+) schema.ShardSetFlushTimes {
+	proto := schema.ShardSetFlushTimes{
+		ByShard: make(map[uint32]*schema.ShardFlushTimes, defaultInitialFlushTimesCapacity),
+	}
+	for _, bucket := range buckets {
```

Comment:
I'd rather keep it simple given that there's nothing that's time-consuming in this operation, which probably takes on the order of microseconds and is not the bottleneck. Also the lock is not heavily contended (the write lock is only needed when a new list is registered, which only happens when the server sees a new policy, which happens relatively rarely).